core/src/main/java/com/linecorp/armeria/common/DefaultRequestHeadersBuilder.java

```diff
@@ -0,0 +1,267 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.Map.Entry;
+
+import javax.annotation.Nullable;
+
+final class DefaultRequestHeadersBuilder
+        extends DefaultHttpHeadersBuilder implements RequestHeadersBuilder {
+
+    DefaultRequestHeadersBuilder() {}
+
+    DefaultRequestHeadersBuilder(DefaultRequestHeaders headers) {
+        super(headers);
+    }
+
+    @Override
+    public RequestHeaders build() {
+        final HttpHeadersBase delegate = delegate();
+        if (delegate != null) {
+            checkState(delegate.contains(HttpHeaderNames.METHOD), ":method header does not exist.");
+            checkState(delegate.contains(HttpHeaderNames.PATH), ":path header does not exist.");
+            return new DefaultRequestHeaders(promoteDelegate());
+        }
+
+        final HttpHeadersBase parent = parent();
+        if (parent != null) {
+            return (RequestHeaders) parent;
+        }
+
+        // No headers were set.
+        throw new IllegalStateException("must set ':method' and ':path' headers");
+    }
+
+    // Request-specific methods.
+
+    @Override
+    public HttpMethod method() {
+        final HttpHeadersBase getters = getters();
+        checkState(getters != null, ":method header does not exist.");
+        return getters.method();
+    }
+
+    @Override
+    public RequestHeadersBuilder method(HttpMethod method) {
+        setters().method(method);
+        return this;
+    }
+
+    @Override
+    public String path() {
+        final HttpHeadersBase getters = getters();
+        checkState(getters != null, ":path header does not exist.");
+        return getters.path();
+    }
+
+    @Override
+    public RequestHeadersBuilder path(String path) {
+        setters().path(path);
+        return this;
+    }
+
+    @Nullable
+    @Override
+    public String scheme() {
+        final HttpHeadersBase getters = getters();
+        return getters != null ? getters.scheme() : null;
+    }
+
+    @Override
+    public RequestHeadersBuilder scheme(String scheme) {
+        setters().scheme(scheme);
+        return this;
+    }
+
+    @Nullable
+    @Override
+    public String authority() {
+        final HttpHeadersBase getters = getters();
+        return getters != null ? getters.authority() : null;
+    }
+
+    @Override
+    public RequestHeadersBuilder authority(String authority) {
+        setters().authority(authority);
+        return this;
+    }
+
+    // Override the return type of the chaining methods in the superclass.
```

Comment:
Not sure if it's strictly better, but a lot of these type of classes accept a template parameter in the base class with the subclass type, so the base class can cast to the template parameter type without having to do so in all children.

Reply:
That's a good idea. Fixed by adding `AbstractHttpHeadersBuilder`. I didn't touch the interfaces however because it looks prettier that way when Javadoc is rendered.